File: Python3/1710_maximum_units_on_a_truck.py
# https://leetcode.cn/problems/maximum-units-on-a-truck/

class Solution:
    def maximumUnits(self, boxTypes, truckSize):
        """
        Inputs:
            boxTypes: list[list[int]]
            truckSize: int
        Outputs:
            res: int
        """
        boxTypes = sorted(boxTypes, key=lambda x: -x[1])
        res = 0
        box = 0
        while truckSize > 0 and box < len(boxTypes):
            num = min(truckSize, boxTypes[box][0])
            res += (num * boxTypes[box][1])
            truckSize -= num
            box += 1
        return res

File: Cpp/1710_maximum_units_on_a_truck.cpp
// https://leetcode.cn/problems/maximum-units-on-a-truck/

class Solution {
public:
    int maximumUnits(vector<vector<int>>& boxTypes, int truckSize) {
        sort(boxTypes.begin(), boxTypes.end(),
            [](const vector<int>& b1, const vector<int>& b2) {
                return b1[1] > b2[1];
            });

        int res = 0, box = 0;
        while ((truckSize > 0) && (box < boxTypes.size())) {
            int num = boxTypes[box][0];
            if (truckSize < num) {
                num = truckSize;
            }
            res += (num * boxTypes[box][1]);
            truckSize -= num;
            box += 1;
        }
        return res;
    }
};
